### DISCUSSION TITLE
Can the ImportSTL function shift axis on data?

### DISCUSSION THREAD

**Discussion by tabarr** [thliebig/openEMS-Project#86] [Q&A Usage]
Having problems with the Hyperlynx export function on my pcb design tools so I am trying another way to bring in my pcb. I have separate stl files for each metal layer. Right now they end up on top of each other when I bring them into CSXcad. Is there a way to offset the data when I bring it in or do I need to move the axis before I create the stl file?

**Comment by LubomirJagos** [comment]
I am using FreeCAD and I use to export each object what I want as separate STL file and simply load them in .m file for openEMS and they are at right position. I think that there is something wrong with your export, maybe you can try to import your object into other 3D software like FreeCAD, Blender, ... if result is same as in CSXcad then your export is done wrong.

**Reply by tabarr** [reply]
I have a .stl file with all 4 board metal layers in it, but am not sure how to get substrate in between the layers. So I made 4 .stl files, one for each metal layer. I guess I could try to shift each layer in freeCAD before exporting to .stl and see what that does.

**Reply by thliebig** [reply]
Usually all objects can have transformations (translations, rotations, scaling etc.) assigned  to them... Have a look [here](https://docs.openems.de/python/CSXCAD/CSPrimitives/CSPrimitives.html#CSXCAD.CSPrimitives.CSPrimitives.AddTransform) maybe...

**Reply by LubomirJagos** [reply]
yes like you're writing, you need to separate things and then assign them under material, here is code snippet what I used to do
```matlab
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% MATERIALS AND GEOMETRY
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
CSX = AddMetal( CSX, 'PEC' );

%% MATERIAL - FR4
CSX = AddMaterial(CSX, 'FR4');
CSX = SetMaterialProperty(CSX, 'FR4', 'Epsilon', 4.3, 'Mue', 1, 'Kappa', 0.000586);
CSX = ImportSTL(CSX, 'FR4', 9600, [currDir '/substrate_gen_model.stl'], 'Transform', {'Scale', fc_unit/unit});

%% MATERIAL - air
CSX = AddMaterial(CSX, 'air');
CSX = SetMaterialProperty(CSX, 'air', 'Epsilon', 1, 'Mue', 1);
CSX = ImportSTL(CSX, 'air', 9300, [currDir '/airBox_gen_model.stl'], 'Transform', {'Scale', fc_unit/unit});

%% MATERIAL - copper
CSX = AddMetal(CSX, 'copper');
CSX = ImportSTL(CSX, 'copper', 9700, [currDir '/top gnd patch_gen_model.stl'], 'Transform', {'Scale', fc_unit/unit});
CSX = ImportSTL(CSX, 'copper', 9800, [currDir '/antenna patch_gen_model.stl'], 'Transform', {'Scale', fc_unit/unit});
CSX = ImportSTL(CSX, 'copper', 9500, [currDir '/bottom gnd patch_gen_model.stl'], 'Transform', {'Scale', fc_unit/unit});
```

this is python equivalent:
```python
#######################################################################################################################################
# MATERIALS AND GEOMETRY
#######################################################################################################################################
PEC = CSX.AddMetal('PEC');

## MATERIAL - FR4
material_1 = CSX.AddMaterial('FR4');
material_1.SetMaterialProperty(epsilon=4.3, mue=1, kappa=0.000586)
material_1.AddPolyhedronReader(os.path.join(currDir,'substrate_gen_model.stl'), priority=9600).ReadFile()

## MATERIAL - air
material_2 = CSX.AddMaterial('air');
material_2.SetMaterialProperty(epsilon=1, mue=1)
material_2.AddPolyhedronReader(os.path.join(currDir,'airBox_gen_model.stl'), priority=9300).ReadFile()

## MATERIAL - copper
material_3 = CSX.AddMetal('copper');
material_3.AddPolyhedronReader(os.path.join(currDir,'top gnd patch_gen_model.stl'), priority=9700).ReadFile()
material_3.AddPolyhedronReader(os.path.join(currDir,'antenna patch_gen_model.stl'), priority=9800).ReadFile()
material_3.AddPolyhedronReader(os.path.join(currDir,'bottom gnd patch_gen_model.stl'), priority=9500).ReadFile()

## MATERIAL - PEC
material_4 = CSX.AddMetal('PEC');
```

**Reply by tabarr** [reply]
I was able to figure out how to move my layers in FreeCAD. So I solved that issue
When I import, I get a lot of messages like this:

Polyhedron_Builder::operator(): Face 143686: Trying reverse order... failed
CSPrimPolyhedron::BuildTree: Warning, found polyhedron has invalud faces and is not a closed surface, setting to 3D solid anyway!

The layers seem to look ok in CSXCAD though, but I might try to simplify the layers and just export the major copper planes and the antenna circuit. 

I am going to try an make .stl files for the fr4 as well. Just need to figure out how to thin the board object in FreeCAD. Then I will try the code snippet.
